### DISCUSSION TITLE
External / Wireless Thermistor replacing native

### DISCUSSION THREAD

**Discussion by serimi89** [revk/xESP32-Faikout#244] [Ideas]
Hey revk and all,
I am new to the faikin project and I am loving it (not new to IoT or AC haha) - thanks so much!

I have 2 different set of Split ACs in my houses, both working different in terms of user interaction. 
- Concealed floor standing (FNA9), controlled via Madoka Wall Remote (BRC1H5x, connected through P1/P2 port), can be set to use the internal thermistor OR the Madoka Wall Remote thermistor using the Field Settings (installer menu).
- Classic wall mount (FTXMR), controlled via IR remote, uses internal thermistor for temperature control.

I am happy with the way the first set up is working and I won't use faikin for this (ok maybe, at some point, just because its nice 😃 ), but I can connect to the Madoka via BLE and therefore control it via HA.

With regards to the second unit, I am not satisfied with the way temperature control works natively (I think widely discussed, the location of the unit itself, high up a wall, in combination with the thermistor placement make no sense). Faikin to the rescue! My idea was to implement a BLE temp sensor (directly or via HA) to have the AC do its thing more efficiently. As far as I understand, at the moment the Faikin software is not able to replace the thermistor values as such, in software, correct? 

So I wanted to share two ideas:

1. There is a wired remote for most of the wall mount units, I believe most units use the BRC073A1. While I dont have one of those units myself, I am under the impression (pictures available online) that it has a dedicated thermistor and it communicates with the main unit via S21 (and through adapters, depending on the unit). Do you think there are some kind of Field Settings (installer menu) just like in the Madoka that could be addressed through the S21 port and using the Faikin? Or any other way they might be able to overwrite the internal thermistor? 
2. Ok maybe this is going to be a bit more effort, but in theory the Faikin could also act as a thermistor by PWM output. Of course this needs to be investigated with regards to accuracy and hardware however afaik there is a lot of data-tables out there showing Daikin Thermistor Resistance values.

Just putting this out here - I am happy to help, investigate re software / hardware. Of course only if this is even relevant to others.

Have a good one!
Seb

**Comment by revk** [comment]
Basically we have found no S21 to tell the Daikin the reference temperature. If you find one, tell me.

I have a "Faikin auto" mode that can work with an external reference including several BLE devices. Fooling the aircon to play out game.

**Reply by serimi89** [reply]
Could get one of those BRC073A1 to check?

**Reply by njh** [reply]
It is quite hard to find information hidden away in threads on these discussion pages.

There is some information about the BRC073 here:
https://github.com/revk/ESP32-Faikin/discussions/226#discussioncomment-8394943

Summary:
- I bought a BRC073 to see if there was a S21 message for reporting indoor temperature from a Wall Contoller
- When I received it, I was happy to see a thermistor in the corner, solder to the board
- However before even inspecting the the S21 messages, I was unable to get it to report the temperate of the thermistor to the indoor unit
- Maybe there is a Field Setting for this but I have not been able to find one
- I have successfully got the BRC073 to talk to the Faikin S21 Simulator, but it has not revealed anything new

Maybe I should write it up properly somewhere.

**Reply by MUN0X** [reply]
I’ve been reading a bit on brc073. It appears you need to enable the external thermistor in the field settings. Have you done that?

Edit: found the answer in the other topic… any luck with reversing the field codes that are available?

**Reply by revk** [reply]
My field settings (not sure what model, ducted unit) had field settings for the P1/P2 based remote, and it does not work (tested with freezer spray on its sensor). Just in case that helps.

**Reply by MUN0X** [reply]
I see that brc073 is wired up with 5 wires. Is it possible that the thermistors are hardwired between 5v and ground directly to the aircon?

**Comment by serimi89** [comment]
> I see that brc073 is wired up with 5 wires. Is it possible that the thermistors are hardwired between 5v and ground directly to the aircon?

You could try and measure things with a multimeter. Could try to follow the traces of the thermistor and see where it ends up. Most likely the processing happens on the brc073, as the empty pin on the S21 pin is 5V.
 
I only have a BRC1E53C which is P1/P2 but judging from photos of the brc073 the PCB looks very similar, just with different connectors (S21 vs P1/P2) - I will have a look, too. 

@revk as far as I understand all values that have a number format are know, right? 
RH - Home
RI - Liquid
Ra - Outside 
RC - Set Temp
RM - Horizontal swing (value is +125)
RN - Vertical swing (value is +125)
RX - Adjusted Set Temp (in my case seems to be RC + 2.5)

**Reply by serimi89** [reply]
Checked the BRC1E53C which is P1/P2 and the thermistor is going into the main processor under the screen, anyone with a brc073 to verify this? 

![IMG_7701](https://github.com/revk/ESP32-Faikin/assets/74360627/f5238a1b-d70f-44a8-972b-55b663d87343)

**Reply by Ambass** [reply]
Hi, 
I'm new here, and already fight a lot with Daikin temperature regulation.

I fully agree, the basis regulation of Daikin is pure shit due to internal sensor placement.
I have 7 stylish units and 4 more basic WiFi connected locally.
In each one I added an ESP32 to force temperature from a external Mi BLE sensor. For stylish , it's quite easy, the sensor is I2C connected.
For others, it's a CTN, so I fake it with the DAC of ESP and resistors network.
The result is quite perfect, regulation ripple is lower than 0.5°C, fan speed remains quiet.

Recently I added a new stylish unit "rev C", and there is no way to connect WiFi without Daikin Cloud Onecta. This is why I tried Faikin !
This is great , but I dicovered the BLE sensor is not injected through S21 port and "auto mode" is not so smooth as faking the sensor itself.

I've to check again, but it seems the ground is common ( at least for stylish ) on S21 and I2C sensor, so it could be possible to use same ESP for Faikin software and  faking I2C sensor

**Reply by serimi89** [reply]
Hey Ambass,

thanks for your comment! Could you elaborate please how you fake a thermistor using the DAC of the ESP and resistors network? Do you have a schematic?

**Reply by nicolas5000** [reply]
Hi @Ambass I have Stylish too from 2023 and it seems that humidity and temp sensors are on the same board. Can you explain how you replaced the I2C temp sensor?
If possible on my units I will try to add the fake sensor to Faikout and propose a PR. Thanks!

**Reply by Ambass** [reply]
Hi @nicolas5000 , 
Yes it's a single I2C sensor with both measurements on a small board ( HSHCAL101B ), you'll see it on pictures..
I disconnect sensor and replaced it by ESP32 module using I2C. 
Temperature and humity are received from the external bluetooth thermometer and sent to I2C. 
Additionally, I also connect sensor on the 2nd I2C line to have recovery values in case of failure of bluetooth one.
Finally the regulation is done smoothly by the aircon unit and very fine (+/- 0.1...0.2°C ), and moreover without playing with the setpoint and fan.
I join pictures ( it's not so nice as the faikin one ! ) and schematic.
To fully complete the job, aircon WiFi board using crappy cloud was removed, and ESP32 provides WiFi access with MQTT throug S21 port.

![IMG_6884](https://github.com/user-attachments/assets/e31e5f6b-60dd-4c07-a5f9-4fd6332a26d7)
![IMG_6885](https://github.com/user-attachments/assets/c6fa34d5-8267-41f7-9ef8-e5fa02144037)

<img width="1696" height="893" alt="image" src="https://github.com/user-attachments/assets/e1edf107-7560-451c-99f9-6985864cb051" />

**Reply by nicolas5000** [reply]
Hi @Ambass , thank you very much for sharing this!
I have all the components ESP32-C6 dev kit, 4 lines 3.3/5V converter, JST PAP... except J1 and J2 connectors. So I think I will do it also!
On my side I would like to try with a Zigbee sensor that should have 0.2°C precision and 5 seconds reactivity from specification (that's why I took ESP32-C6, but I think about ESP32-C5 too because it have extra RAM like S3).

Did you modify Faikout source code to add this or is it your project including S21 management from Faikout?

**Reply by njh** [reply]
@Ambass this is really impressive! Well done!

**Reply by Ambass** [reply]
Hi @nicolas5000 ,
Connectors are available here : https://www.amazon.fr/dp/B0CRZ9D6MQ?ref_=ppx_hzsearch_conn_dt_b_fed_asin_title_4&th=1
It's not Faikout project, it's my own. It's shared with other home automated devices. I only picked up S21 protocol info on Faikout and other threads ( And unfortunately, I never found any way to enter temperature ! )
If you want, I can extract the I2C part ( I use VScode & platformIO ). It could help you to avoid using oscilloscope to debug I2C !

**Reply by nicolas5000** [reply]
Thanks @Ambass , if fact I found that I bought some of these connectors in 2021, so I have everything to work on this!
I use to work with VSCode too, if you can share the I2C part of your project I am very interested, as I will not have to use my digital analyzer to capture the I2C exchanges to reproduce the behaviour!

**Reply by Ambass** [reply]
Hi @nicolas5000 ,
I attach .c file and datasheet of the sensor. There is only a call at initialisation and one every 1 sec.
Fore sure you will find naming/comments in a french/english mixture !!!
Keep in mind an offset of +4°C should be applied in heat mode and -1.5°C in cool mode. ( A poor compensation of daikin due to the weird sensor position ). This offset is constant in all temperature range.
For your tests, as soon as the unit do not get correct information from ESP, it turns off with the power light blinking.
The schematic has been done later, in case of doubt, trust pictures ( They are working ! )

[clim_i2c.zip](https://github.com/user-attachments/files/23913719/clim_i2c.zip)
[hshcal101b_data.pdf](https://github.com/user-attachments/files/23913728/hshcal101b_data.pdf)

**Reply by nicolas5000** [reply]
Thanks a lot @Ambass !
I found in some Daikin service manual that there is a +2.5°C offset to the value and the unit stops heating at an additional +2.0°C. I think it is closed to the +4°C offset you found (if you ask for 20°C target it heats until 24.5°C but displays 24°C, so +4°C)

I will take time to create the hardware part, probably this week, and test the original Faikout project on my ESP32-C6.
Then I will integrate and test "I2C remote sensor" as a new feature in Faikout project for everyone having the I2C sensor. It could take some time but I think it will be so easier with your data!
Don't worry about comments in "french", I'm also from France :smiley:

**Comment by Ambass** [comment]
Hi , 
I've this :
As you can see on pictures, CTN wire is cut on PCB to obtain schematic.
This is working for values from 15° to 30°C ( 0 to 3300mv out from DAC )
 
    float tb_temp[8]    = { 0, 5, 10, 15, 20, 25, 30, 35 } ;  // temperature breakpoints
    float tb_ohm[8]     = { 26851, 21799, 17798, 14612, 12058, 10000, 8332.5, 6974.6 } ; // CTN values
    float tb_out_mv[8]  = { 0, 0, 0, 870, 1700, 2490, 3240, 3300 } ; // DAC output

![ctn](https://github.com/user-attachments/assets/baa97738-00b9-431f-9f23-8f4f9b357beb)
![IMG_4854](https://github.com/user-attachments/assets/b9b80a1e-6acd-4e34-98ad-4f9fc180b0ca)
![IMG_4855](https://github.com/user-attachments/assets/b11b7467-82d0-48c9-bbf4-55f0e20f601a)
![IMG_4856](https://github.com/user-attachments/assets/7ad35519-0543-41df-ae0a-c606354af092)

**Reply by joealden** [reply]
@Ambass sorry to hijack and be off topic (let me know if there is a better place to talk if you are willing), but I'm reaching out to Stylish owners to try to figure out if a noise coming from my new Stylish units is expected - see https://github.com/revk/ESP32-Faikin/discussions/425#discussioncomment-11339496 - thanks!
